src/transport.rs

```diff
@@ -0,0 +1,279 @@
+                thread::sleep( SWI_BIT_SEND_DELAY * 8);
+            },
+        }
+    }
+
+    pub fn send_recv_buf(&mut self, delay: Duration, buf: &mut BytesMut) -> Result {
+        match self.protocol {
+            TransportProtocol::I2c => {
+                self.send_i2c_buf(self.i2c_address, &buf[..])?;
+                thread::sleep(delay);
+                self.recv_i2c_buf(buf)
+            },
+            TransportProtocol::Swi => {
+                let _ = self.swi_port.as_mut().unwrap().clear(ClearBuffer::All);
+                let swi_msg = self.encode_uart_to_swi(buf);
+                self.send_swi_buf(&swi_msg)?;
+                thread::sleep(delay);
+                self.recv_swi_buf(buf)
+            },
+        }
+    }
+
+    fn send_i2c_buf(&mut self, address: u16, buf: &[u8]) -> Result {
+        let write_msg = i2c_linux::Message::Write {
+            address,
+            data: buf,
+            flags: Default::default(),
+        };
+
+        self.i2c_port.as_mut().unwrap().i2c_transfer(&mut [write_msg])?;
+        Ok(())
+    }
+
+    fn recv_i2c_buf(&mut self, buf: &mut BytesMut) -> Result {
+        unsafe { buf.set_len(1) };
+        buf[0] = 0xff;
+        for _retry in 0..RECV_RETRIES {
+            let msg = i2c_linux::Message::Read {
+                address: self.i2c_address,
+                data: &mut buf[0..1],
+                flags: Default::default(),
+            };
+            if let Err(_err) = self.i2c_port.as_mut().unwrap().i2c_transfer(&mut [msg]) {
+            } else {
+                break;
+            }
+            thread::sleep(RECV_RETRY_WAIT);
+        }
+        let count = buf[0] as usize;
+        if count == 0xff {
+            return Err(Error::timeout());
+        }
+        unsafe { buf.set_len(count) };
+        let read_msg = i2c_linux::Message::Read {
+            address: self.i2c_address,
+            data: &mut buf[1..count],
+            flags: ReadFlags::NO_START,
+        };
+        self.i2c_port.as_mut().unwrap().i2c_transfer(&mut [read_msg])?;
+        Ok(())
+    }
+
+    fn send_swi_buf(&mut self, buf: &[u8]) -> Result {
+        
+        let send_size = self.swi_port.as_mut().unwrap().write(buf)?;
+
+        //Each byte takes ~45us to transmit, so we must wait for the transmission to finish before proceeding
+        let uart_tx_time = buf.len() as u32 * SWI_BIT_SEND_DELAY; 
+        thread::sleep(uart_tx_time);
+        //Because Tx line is linked with Rx line, all sent msgs are returned on the Rx line and must be cleared from the buffer
+        let mut clear_rx_line = BytesMut::new();
+        clear_rx_line.resize(send_size, 0);
+        let _ = self.swi_port.as_mut().unwrap().read_exact( &mut clear_rx_line );
+
+        Ok(())
+    }
+
+    fn recv_swi_buf(&mut self, buf: &mut BytesMut) -> Result {
+        unsafe { buf.set_len(2) };
+        buf[1] = 0xFF;
+
+        let mut transmit_flag = BytesMut::new();
```

Comment:
Recommend not using BytesMut for simple known sized arrays or single bytes as I think is the case here.